java/google/registry/rde/XjcToDomainResourceConverter.java

```diff
@@ -0,0 +1,329 @@
+import google.registry.xjc.domain.XjcDomainStatusType;
+import google.registry.xjc.rdedomain.XjcRdeDomain;
+import google.registry.xjc.rdedomain.XjcRdeDomainElement;
+import google.registry.xjc.rdedomain.XjcRdeDomainTransferDataType;
+import google.registry.xjc.rgp.XjcRgpStatusType;
+import google.registry.xjc.secdns.XjcSecdnsDsDataType;
+import org.joda.time.DateTime;
+
+/** Utility class that converts an {@link XjcRdeDomainElement} into a {@link DomainResource}. */
+final class XjcToDomainResourceConverter extends XjcToEppResourceConverter {
+
+  private static final XmlToEnumMapper<TransferStatus> TRANSFER_STATUS_MAPPER =
+      XmlToEnumMapper.create(TransferStatus.values());
+
+  private static final Function<XjcDomainStatusType, StatusValue> STATUS_CONVERTER =
+      new Function<XjcDomainStatusType, StatusValue>() {
+    @Override
+    public StatusValue apply(XjcDomainStatusType status) {
+      return convertStatusType(status);
+    }
+  };
+
+  private static final Function<String, Key<HostResource>> HOST_OBJ_CONVERTER =
+      new Function<String, Key<HostResource>>() {
+        @Override
+        public Key<HostResource> apply(String fullyQualifiedHostName) {
+          Key<HostResource> key = ForeignKeyIndex.loadAndGetKey(
+              HostResource.class,
+              fullyQualifiedHostName,
+              DateTime.now());
+          checkState(key != null,
+              String.format("HostResource not found with name '%s'", fullyQualifiedHostName));
+          return key;
+        }
+  };
+
+  private static final Function<XjcDomainContactType, DesignatedContact> CONTACT_CONVERTER =
+      new Function<XjcDomainContactType, DesignatedContact>() {
+        @Override
+        public DesignatedContact apply(XjcDomainContactType contact) {
+          return convertContactType(contact);
+        }
+  };
+
+  private static final Function<XjcSecdnsDsDataType, DelegationSignerData> SECDNS_CONVERTER =
+      new Function<XjcSecdnsDsDataType, DelegationSignerData>() {
+
+        @Override
+        public DelegationSignerData apply(XjcSecdnsDsDataType secdns) {
+          return convertSecdnsDsDataType(secdns);
+        }
+
+  };
+
+  /** Converts {@link XjcRgpStatus} to {@link GracePeriod} */
+  private static class GracePeriodConverter
+      implements Function<XjcRgpStatusType, GracePeriod> {
+
+    private final XjcRdeDomain domain;
+    private final Key<BillingEvent.Recurring> autoRenewBillingEvent;
+    private final Registry tld;
+
+    GracePeriodConverter(XjcRdeDomain domain, Key<BillingEvent.Recurring> autoRenewBillingEvent) {
+      this.domain = domain;
+      this.autoRenewBillingEvent = autoRenewBillingEvent;
+      this.tld = Registry.get(
+          Registries.findTldForNameOrThrow(InternetDomainName.from(domain.getName())).toString());
+    }
+
+    @Override
+    public GracePeriod apply(XjcRgpStatusType gracePeriodStatus) {
+      // TODO: (wolfgang) address these items in code review:
```

Comment:
Wolfgang, I am going to look into the grace period code. Can you give me a sense of your strategy regarding billing events? You are not creating one-time billing events, which makes sense in one way, since we didn't bill anyone. But I will have to check into whether this causes problems elsewhere. Has there already been a discussion about this issue that you can point me to? Sorry, just trying to get up to speed.

P.S. You are a mighty warrior indeed for writing all this import code.